enhancements/ingress/aws-load-balancer-operator.md

```diff
@@ -0,0 +1,384 @@
+specify which features/addons have to be disabled. By default all the addons
+will be enabled. They are disabled by setting the addon CLI flag to false. When
+an addon which was previously enabled is disabled the controller does not remove
+the existing addon attachment from the provisioned load balancers.
+
+The operator will also create a
+[CredentialRequest](https://docs.openshift.com/container-platform/4.9/rest_api/security_apis/credentialsrequest-cloudcredential-openshift-io-v1.html#credentialsrequest-cloudcredential-openshift-io-v1)
+on behalf of the lb-controller and mount the minted credentials in the
+_Deployment_ of the controller. This means that the operator will only work on
+OCP/OKD but this is sufficient for the initial release.
+
+
+### Risks and Mitigations
+
+#### Restricting features in the lb-controller
+
+Some changes will have to be made in the lb-controller to ensure that the users
+only use features which are supported on OCP and only those features which
+target Ingress resources. These changes may also be contributed back upstream
+since it may benefit other users.
+
+1. The annotation `alb.ingress.kubernetes.io/target-type` can be set to the
+value **_ip_** or **_instance_**. When the value is set to _instance_ the
+created load balancer will be configured to route traffic through a NodePort
+service. The _ip_ value only works when the cluster uses the Amazon VPC CNI and
+hence will not work on OCP/OKD clusters. The controller has to be modified so
+that this annotation is ignored and the controller always uses routing through
+_NodePort_ services.
+2. The aws-load-balancer-controller has support for _Service_ type resources as
+well. When the annotation `service.beta.kubernetes.io/aws-load-balancer-type:
+"external"` is specified on a _Service_ resource of type _LoadBalancer_ the
+in-tree controller ignores this resource and the lb-controller instead
+provisions a Network Load Balancer with the correct configuration. Since we want
+to restrict the features supported by the controller to only Ingresses the
+lb-controller will have to be modified so that it does not reconcile any Service
+resources.
+
+#### Parallel operation  of the Openshift Router and lb-controller
+
+The openshift Router manages ingresses which don’t have any ingress class value.
+This is the expected behavior currently and the lb-controller should not attempt
+to reconcile Ingresses which don’t have an ingress class set. If the user wants
+to switch over to using the lb-controller as default they can create an
+_IngressClass_ resource with the annotation
+`ingressclass.kubernetes.io/is-default-class` set to _true_ and the
+`spec.controller` value should be _ingress.k8s.aws/alb._ The user can also
+create multiple `IngressClass` resources with different configurations and have
+the `spec.controller` set to _ingress.k8s.aws/alb._ All Ingresses which belong
+to these classes will be reconciled by the lb-controller.
```

Comment:
How is this possible? `aws-load-balancer-controller` expects only 1 `IngressClass`, so will reconcile only `Ingress`es of this class.

Reply:
Because that's how it's been implemented in the controller. The CLI parameter is kind of non-obvious in how it works. It the ingress class is specified through the annotation then it limits it to that parameter. However if it is specified through the field in the spec then it ignores that parameter and only verifies that the corresponding `IngressClass` has the `controller` field set to the value mentioned above. Now that I think about it might be confusing for the user. We should probably disable the annotations based Ingress Class altogether.

Reply:
Here's the how the ingress is [classified](https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/46668f11bdc548a3df60d1a921b0e70cbf0f91f1/pkg/ingress/group_loader.go#L171). Maybe we should hide the `--ingress-class` parameter altogether from the user. And expect the user to create the `IngressClass` manually.

Reply:
Interesting. Yes, that kinda makes sense, as now, with `IngressClass` resource there is no need to distinguish by the class name (like in the annotation), `.spec.controller` came into play and many classes can now be implemented by the same controller which creates a new abstraction layer.
I think that the docs are not well updated in some places cause I definitely saw the statements about `--ingress-class` flag limiting the reconciliation logic. Also, I'm still confused about how multiple `IngressClassParams` with conflicting fields are managed.

Anyway, taking this logic into account, I think I agree with you that setting `--disable-ingress-class-annotation` as default and removing `IngressClass` from the API is a nice option.

Reply:
Just to finish on this one. I don't know why I thought that multiple `IngressClassParams` would cause a conflict, a brain bug I guess. There is no problem with this as each ingress belongs to only 1 ingress class. The conflict can be only if ingresse belonging to different classes are grouped in a single ingress group, in that case different fields from the `IngressClassParams` may cause the conflict, like here when different schemes are used:
```
{"level":"debug","ts":1643651687.9120367,"logger":"controller-runtime.manager.events","msg":"Warning","object":{"kind":"Ingress","namespace":"hello2","name":"hello-openshift","uid":"0e68ba57-39ea-49a4-af63-90fe1034288b","apiVersion":"networking.k8s.io/v1","resourceVersion":"167518"},"reason":"FailedBuildModel","message":"Failed build model due to conflicting scheme: map[internal:{} internet-facing:{}]"}
```
But this is no different from the situation when a single ingress class is used and multiple ingresses set conflicting settings in the annotations.